### PR TITLE
feat: type annotations

### LINKED ISSUE #27: Type annotations
Currently we have no annotations at all. Since most parameters and return are `str`, it is fast to fix it.

If it is needed, I'll give the pr.

### PR DESCRIPTION
**Pull request by Master-Hash**
Fix #27

### CODE REVIEW

File: ada_url/ada_adapter.py [hunk 1]
```diff
@@ -290,7 +320,9 @@ def normalize_url(s):
     return parse_url(s, attributes=('href',))['href']
 
 
-def parse_url(s, attributes=PARSE_ATTRIBUTES):
+# FIXME constrain `attributes``
```
Comment (anonrig) [inline ada_url/ada_adapter.py:323]:
```suggestion
# FIXME constrain `attributes`
```

**Review by anonrig — COMMENTED** [review]
Looks good to me, although @bbayles probably has more knowledge on this than me.

File: ada_url/ada_adapter.py [hunk 2]
```diff
@@ -354,7 +386,8 @@ def parse_url(s, attributes=PARSE_ATTRIBUTES):
     return ret
 
 
-def replace_url(s, **kwargs):
+# FIXME constrain key of `**kwargs` by `URL_ATTRIBUTES`
```
Comment (bbayles) [inline ada_url/ada_adapter.py:389]: The docstring says that unknown keys are ignored; I'd like to preserve this behavior.

File: ada_url/ada_adapter.py (hunk 1 above)
Comment (bbayles) [reply ada_url/ada_adapter.py:323]: Same here - the docstring says unrecognized requests are ignored; we need not constrain with the type if we preserve that behavior.

File: ada_url/ada_adapter.py (hunk 2 above)
Comment (Master-Hash) [reply ada_url/ada_adapter.py:389]: I see, but typing isn't just about error checking; constraints make auto suggestions work better.

If you insist, I'll just delete the comment; or I'll use `Union[Literal["href"], ...]`.

Comment (bbayles) [reply ada_url/ada_adapter.py:389]: Let's remove the comments and create a GitHub issue to help make auto suggestions (what editor?) better.

**Review by bbayles — APPROVED** [review]

**Comment by bbayles** [comment]
I'd like to add tests of the annotations with mypy, but we can do that in another PR.

**Comment by anonrig** [comment]
🚀

**Pull request merged**